tensorlayer/layers/pooling.py

```diff
@@ -249,19 +266,214 @@ def meanpool3d(net, filter_size=(3, 3, 3), strides=(2, 2, 2), padding='valid', d
+
+        # the input of this layer is the output of previous layer (fixed)
+        self.inputs = layer.outputs
+
+        # print out info (customized)
+        logging.info("GlobalMeanPool1d %s" % name)
+
+        # operation (customized)
+        self.outputs = tf.reduce_mean(layer.outputs, axis=1, name=name)
+
+        # get stuff from previous layer (fixed)
+        self.all_layers = list(layer.all_layers)
+        self.all_params = list(layer.all_params)
+        self.all_drop = dict(layer.all_drop)
+
+        # update layer (customized)
+        self.all_layers.append(self.outputs)
+        # self.all_params.extend( [W, b] )
+
+
+class GlobalMaxPool2d(Layer):
+    """ The :class:`GlobalMaxPool2d` class is a 2D Global Max Pooling layer.
+
+    Parameters
+    ------------
+    layer : :class:`Layer`
+        The previous layer with a output rank as 4.
+    name : str
+        A unique layer name.
+
+    Examples
+    ---------
+    >>> x = tf.placeholder("float32", [None, 100, 100, 30])
+    >>> n = InputLayer(x, name='in2')
+    >>> n = GlobalMaxPool2d(n)
+    ... [None, 30]
+    """
+
+    def __init__(
+            self,
+            layer=None,
+            name='globalmaxpool2d',
+    ):
+        # check layer name (fixed)
+        Layer.__init__(self, name=name)
+
+        # the input of this layer is the output of previous layer (fixed)
+        self.inputs = layer.outputs
+
+        # print out info (customized)
+        logging.info("GlobalMaxPool2d %s" % name)
+
+        # operation (customized)
+        self.outputs = tf.reduce_max(layer.outputs, axis=[1, 2], name=name)
+
+        # get stuff from previous layer (fixed)
+        self.all_layers = list(layer.all_layers)
+        self.all_params = list(layer.all_params)
+        self.all_drop = dict(layer.all_drop)
+
+        # update layer (customized)
+        self.all_layers.append(self.outputs)
+        # self.all_params.extend( [W, b] )
+
+
+class GlobalMeanPool2d(Layer):
+    """ The :class:`GlobalMeanPool2d` class is a 2D Global Mean Pooling layer.
+
+    Parameters
+    ------------
+    layer : :class:`Layer`
+        The previous layer with a output rank as 4.
+    name : str
+        A unique layer name.
+
+    Examples
+    ---------
+    >>> x = tf.placeholder("float32", [None, 100, 100, 30])
+    >>> n = InputLayer(x, name='in2')
+    >>> n = GlobalMeanPool2d(n)
+    ... [None, 30]
+    """
+
+    def __init__(
+            self,
+            layer=None,
+            name='globalmeanpool2d',
+    ):
+        # check layer name (fixed)
+        Layer.__init__(self, name=name)
+
+        # the input of this layer is the output of previous layer (fixed)
+        self.inputs = layer.outputs
+
+        # print out info (customized)
+        logging.info("GlobalMeanPool2d %s" % name)
+
+        # operation (customized)
+        self.outputs = tf.reduce_mean(layer.outputs, axis=[1, 2], name=name)
+
+        # get stuff from previous layer (fixed)
```

Comment:
@luomai shall we consider slowly move this into the `Layer` base class?

```
 self.all_layers = list(layer.all_layers)
 self.all_params = list(layer.all_params)
 self.all_drop = dict(layer.all_drop)
```

Reply:
in `ALayer(Layer):`
        Layer.__init__(self, name=name, layer=layer)
in `Layer`:
```
    class Layer(object):
        def __init__(self, name, layer):
             self.all_layers = list(layer.all_layers)
            self.all_params = list(layer.all_params)
            self.all_drop = dict(layer.all_drop)
```

so we don't have to do it every layer.

Reply:
We can change it slowly by:
1. introduce `layer` argument to `__init__` of `Layer` as optional.
2. slowly make all sub layers pass the layer argument to `Layer.__init___`
3. make `layer` required in `__init__`

Reply:
we can do this in another repo?

Reply:
Of course, this needs a careful plan before carrying out.